decode_test.go

```diff
@@ -1715,6 +1715,58 @@ func (s *S) TestUnmarshalKnownFields(c *C) {
 	}
 }
 
+func (s *S) TestDecoderHook(c *C) {
+	y := `
+a:
+  b: foo
+  c:
+    d:
+      e: bar
+f:
+  g:
+    h:
+      i:
+        - 1
+        - 2
+j: {"k": true, "l": false}
+
+x:
+  - z: yes
+    y: no
+  - z: up
+    w: down
+`
```

Comment:
If I replace this with `"Bare document"`, the hook is not called. I found that a bit surprising. WDYT about that?

Reply:
Hmm, well the hook is called for every "leaf field", so there has to be a yaml map for the hook to be called. Just a string is only a value and not a map, so from that perspective it's working as intended.

Btw, this was added specifically so we can lint the document without parsing it into an intermediary representation like a `map[string]interface{}`. Changing the hook so it's called in case the document is a simple value wouldn't give us any benefit as far as I can see.